### DISCUSSION TITLE
Ability to drag the devtools icon

### DISCUSSION THREAD

**Discussion by toto1384** [TanStack/query#3794] [Ideas]
Even though the icon is small and in the corner of the screen, it would be nice if we had the ability to drag it whenever we want.

**Comment by TkDodo** [comment]
clicking the button opens the devtools, so how would that work? Having the same icon support click and drag doesn't sound too good...

i think you can already specify where the icon should sit (4 corners of the screen)

**Reply by toto1384** [reply]
You drag it wherever you want on the screen and when you press it, it should open the dev tools just as it did before. It's dx at the end of the day. Maybe you have a part of the UI in the corner of the screen and you don't want to go to the code and specify the position just to see it

**Reply by TkDodo** [reply]
if you know how to do this in an accessible way and so that there is a clear distinction between click and press, feel free to PR it. I think `dnd-kit` is the best tool to do any kind of drag and drop functionality: https://dndkit.com/

**Reply by toto1384** [reply]
I will consider it, thanks for the suggestion

**Comment by fum4** [comment]
I would like to tackle this feature request. I’ve come across a scenario where my app displays a different button or action in every corner of the screen, depending on the active page. These will get covered by the devtools button as long as its position is static.

@TkDodo I have a small POC in place for this, but I think you might want to tackle the following topics first:

1. Position Persistence: Should the button's position be saved and restored between page reloads (e.g., in localStorage)? Or should it reset to the default position each time?

2. Opt-in vs Default: Should draggable functionality be:
- Always enabled?
- Opt-in via a new prop (e.g., draggable={true})?

**Reply by DamianOsipiuk** [reply]
Why not just use [DevtoolsPanel](https://tanstack.com/query/latest/docs/framework/react/devtools#embedded-mode) and position it however you want, with whatever trigger you want?

Also if you are using multiple tanstack libraries you might want to consider switching to [TanStack
Devtools](https://tanstack.com/devtools/latest).
You can also add your own plugins there and have a singular devtools panel for everything you need.

**Reply by fum4** [reply]
I was not aware of [TanStack Devtools](https://tanstack.com/devtools/latest). It looks **great** 👏 I will give it a try as soon as possible, thanks for the tip.

However, we only use TanStack Query at work, at least for the moment. It does look like TanStack Devtools comes with a bunch of other cool stuff that benefits any kind of app, but `query-devtools` are currently enough for our use case.

Either way, I see that TanStack Devtools trigger button has the same limitation, it can only be rendered at a predefined static position (unlike the `react-scan`-like widget), which would cause us the same incovenience (covering actionable parts of the screen).

This can be indeed handled by app code from our side (at least in the case of `query-devtools`), which we already did in the meanwhile, but I think it could be a helpful addition for others too.

The [DevtoolsPanel](https://tanstack.com/query/latest/docs/framework/react/devtools#embedded-mode) might also come handy (if I understand correctly you suggest that devs could implement their own convenient button -- a draggable one in my situation -- that triggers the devtools). However this implies some manual work and the floating mode **already has the ideal DX** imo, except that the trigger button is fixed. So in my case, I would only try to reproduce the floating mode behaviour, using my own draggable button.

@DamianOsipiuk If you think this addition would be helpful however, I have a working POC ready to be reviewed and I am willing to take a look on the TanStack Devtools repo and integrate it there too 👍

Thanks!
